File: src/C/cholesky.c
#include "chompack.h"

int cholesky(const int_t n,         // order of matrix
	     const int_t nsn,       // number of supernodes/cliques
	     const int_t *snpost,   // post-ordering of supernodes
	     const int_t *snptr,    // supernode pointer
	     const int_t *relptr,
	     const int_t *relidx,
	     const int_t *chptr,
	     const int_t *chidx,
	     const int_t *blkptr,
	     double * restrict blkval,
	     double * restrict fws,  // frontal matrix workspace
	     double * restrict upd,  // update matrix workspace
	     int_t * restrict upd_size
	     ) {

  int nn,na,nj,offset,info,i,j,k,ki,l,N,nup=0;
  double * restrict U;
  int iOne=1;
  double dOne=1.0,dNegOne=-1.0;
  char cL='L',cT='T',cR='R',cN='N';

  U = upd;   // pointer to top of update storage

  for (ki=0;ki<nsn;ki++) {
    k = snpost[ki];
    nn = snptr[k+1]-snptr[k];
    na = relptr[k+1]-relptr[k];
    nj = na + nn;

    // build frontal matrix
    dlacpy_(&cL, &nj, &nn, blkval+blkptr[k], &nj, fws, &nj);
    for (j=nn;j<nj;j++) {
      for (i=j;i<nj;i++) {
	fws[nj*j+i] = 0.0; // zero out (2,2) block of frontal matrix
      }
    }

    // add update matrices to frontal matrix
    for (l=chptr[k+1]-1;l>=chptr[k];l--) {
      nup--;
      U -= upd_size[nup]*upd_size[nup];
      // extend-add
      offset = relptr[chidx[l]];
      N = relptr[chidx[l]+1] - offset;
      for (j=0; j<N; j++) {
	for (i=j; i<N; i++) {
	  fws[nj*relidx[offset+j]+relidx[offset+i]] += U[N*j+i];
	}
      }
    }

    // factor L_{Nk,Nk}
    dpotrf_(&cL, &nn, fws, &nj, &info);
    if (info) return info;

    // if supernode k is not a root node, compute and push update matrix onto stack
    if (na > 0) {
      // compute L_{Ak,Nk} := A_{Ak,Nk}*inv(L_{Nk,Nk}')
      dtrsm_(&cR, &cL, &cT, &cN, &na, &nn, &dOne, fws, &nj, fws+nn, &nj);

      // compute Uk = Uk - L_{Ak,Nk}*inv(D_{Nk,Nk})*L_{Ak,Nk}'
      if (nn == 1) {
	dsyr_(&cL, &na, &dNegOne, fws+nn, &iOne, fws+nn*nj+nn, &nj);
      }
      else {
	dsyrk_(&cL, &cN, &na, &nn, &dNegOne, fws+nn, &nj, &dOne, fws+nn*nj+nn, &nj);
      }

      // compute L_{Ak,Nk} := L_{Ak,Nk}*inv(L_{Nk,Nk})
      dtrsm_(&cR, &cL, &cN, &cN, &na, &nn, &dOne, fws, &nj, fws+nn, &nj);

      // copy update matrix to stack
      upd_size[nup++] = na;
      dlacpy_(&cL, &na, &na, fws+nn*nj+nn, &nj, U, &na);
      U += na*na;
    }

    // copy the leading nn columns of frontal matrix to blkval
    dlacpy_(&cL, &nj, &nn, fws, &nj, blkval+blkptr[k], &nj);
  }
  return 0;
}

File: src/python/pybase/cholesky.py
from cvxopt import matrix, blas, lapack
from chompack.symbolic import cspmatrix
from chompack.misc import frontal_add_update

def cholesky(X):
    """
    Supernodal multifrontal Cholesky factorization:

    .. math::
         X = LL^T

    where :math:`L` is lower-triangular. On exit, the argument :math:`X`
    contains the Cholesky factor :math:`L`.

    :param X:    :py:class:`cspmatrix`
    """

    assert isinstance(X, cspmatrix) and X.is_factor is False, "X must be a cspmatrix"

    n = X.symb.n
    snpost = X.symb.snpost
    snptr = X.symb.snptr
    chptr = X.symb.chptr
    chidx = X.symb.chidx

    relptr = X.symb.relptr
    relidx = X.symb.relidx
    blkptr = X.symb.blkptr
    blkval = X.blkval

    stack = []

    for k in snpost:

        nn = snptr[k+1]-snptr[k]       # |Nk|
        na = relptr[k+1]-relptr[k]     # |Ak|
        nj = na + nn                   

        # build frontal matrix
        F = matrix(0.0, (nj, nj))
        lapack.lacpy(blkval, F, offsetA = blkptr[k], m = nj, n = nn, ldA = nj, uplo = 'L')

        # add update matrices from children to frontal matrix
        for i in range(chptr[k+1]-1,chptr[k]-1,-1):
            Ui = stack.pop()
            frontal_add_update(F, Ui, relidx, relptr, chidx[i])

        # factor L_{Nk,Nk}
        lapack.potrf(F, n = nn, ldA = nj)

        # if supernode k is not a root node, compute and push update matrix onto stack
        if na > 0:   
            # compute L_{Ak,Nk} := A_{Ak,Nk}*inv(L_{Nk,Nk}')
            blas.trsm(F, F, m = na, n = nn, ldA = nj, 
                      ldB = nj, offsetB = nn, transA = 'T', side = 'R')

            # compute Uk = Uk - L_{Ak,Nk}*inv(D_{Nk,Nk})*L_{Ak,Nk}'
            if nn == 1:
                blas.syr(F, F, n = na, offsetx = nn, \
                         offsetA = nn*nj+nn, ldA = nj, alpha = -1.0)
            else:
                blas.syrk(F, F, k = nn, n = na, offsetA = nn, ldA = nj,
                          offsetC = nn*nj+nn, ldC = nj, alpha = -1.0, beta = 1.0)

            # compute L_{Ak,Nk} := L_{Ak,Nk}*inv(L_{Nk,Nk})
            blas.trsm(F, F, m = na, n = nn,\
                      ldA = nj, ldB = nj, offsetB = nn, side = 'R')

            # add Uk to stack
            Uk = matrix(0.0,(na,na))
            lapack.lacpy(F, Uk, m = na, n = na, uplo = 'L', offsetA = nn*nj+nn, ldA = nj)
            stack.append(Uk)

        # copy the leading Nk columns of frontal matrix to blkval
        lapack.lacpy(F, blkval, uplo = "L", offsetB = blkptr[k], m = nj, n = nn, ldB = nj)        

    X.is_factor = True

    return

